dev/common/tdd.py

```diff
@@ -0,0 +1,111 @@
+#!/usr/bin/env python3
+
+import json
+import logging
+import os
+import subprocess
+import sys
+
+
+def get_current_branch():
+    pid = subprocess.run(
+        'git branch --show-current',
+        shell=True,
+        stdout=subprocess.PIPE,
+        check=True
+    )
+    branch_name = pid.stdout.decode('utf-8')
+    branch_name = branch_name.strip()
+    return branch_name
+
+
+def get_changed_files(pr_branch, target_branch="master"):
+    cmd = f'git diff --name-only {pr_branch}..{target_branch}'
+    pid = subprocess.run(cmd, shell=True, stdout=subprocess.PIPE, check=True)
+    filenames = pid.stdout.decode('utf-8')
+    filenames = filenames.split('\n')
+    filenames = [x.strip() for x in filenames if x.strip()]
+    return filenames
+
+
+def verify_test_files_changed(changed_files):
+
+    # places where appplication source goes
+    app_paths = [
+        'galaxy_ng/'
+    ]
+
+    # places where test code goes
+    test_paths = [
+        'galaxy_ng/tests'
+    ]
+
+    def is_app_path(fn):
+        for ap in app_paths:
+            if fn.startswith(ap):
+                return True
+        return False
+
+    def is_test_path(fn):
+        for tp in test_paths:
+            if fn.startswith(tp):
+                return True
+        return False
+
+    # exit early if no non-test changed in the api code
+    app_changed = False
+    for cf in changed_files:
+        if is_app_path(cf) and not is_test_path(cf):
+            app_changed = True
+    if not app_changed:
+        return
+
+    # look for any changes to file in the tests dirs
+    tests_found = False
+    for cf in changed_files:
+        for tp in test_paths:
+            if cf.startswith(tp):
+                tests_found = True
+                break
+        if tests_found:
+            break
+
+    if not tests_found:
+        raise Exception('Tests should be added or modified with -every- PR!!!')
```

Comment:
This bit may be overly ambitious :)

Should the policy be that we ignore this check for those kinds of PRs or should we enable this only for ..feature PRs? (Like, only run if `CHANGES/*.feature` is being updated as well.)

(Looking at the last 16 merged PRs, 12 of those should not be touching tests.. CI changes, dev env changes, ephemeral, dependency changes (updating pulp), release PRs .. there's probably more?)

Reply:
The code is already filtering out any changes to files outside the application source path.

If tests aren't written for bugfixes, you will inevitably have to fix the same bug again. There should be no excuse for not writing tests for every PR that changes the application's behavior.

https://evolveum.com/test-driven-bugfixing/#:~:text=You%20write%20an%20(automated)%20test,that%20the%20test%20is%20passing.

Reply:
I agree with @jctanner about 90% and the other 10% is covered by @mdellweg there. Features and bugfixes? Absolutely should get new or updated tests. Refactoring I'd almost say SHOULDN'T have test changes, almost by definition of "refactoring". As long as people are responsible about it, maybe we could add a flag you can put in PR or commit descriptions to mark it as refactoring and accepting the existing test state.

Reply:
Without coverage reports or any sense of what functions get tested, i'd argue that a refactor absolutely needs to add tests to -prove- that the author actually took the time to understand and learn what the code was meant to do.